OSCORE: Link target attribute

rikard-sics opened the pull request:
This PR adds the link target attribute 'osc' for resources which are only accessible using OSCORE.

It also corrects the behavior of the _OSCoreResource_ when it is set to only be accessible using OSCORE, but receives a non-OSCORE request.

rikard-sics commented:
This was another point brought up during the recent interop tests.

This functionality is in accordance with the OSCORE RFC Section 9:
https://datatracker.ietf.org/doc/html/rfc8613#section-9

Not sure if JUnit tests are also desired, if so I can add them.

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/coap/LinkFormat.java:
```diff
@@ -51,6 +51,7 @@ public class LinkFormat {
 	public static final String TITLE = "title";
 	public static final String OBSERVABLE = "obs";
 	public static final String LINK = "href";
+	public static final String OSCOREONLY = "osc";
```
Please add 

```
// RFC 8613
```

rikard-sics replied:
Added in commit 4281708a39346675a1d6ee0f88f26bab0c902280

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/server/resources/ResourceAttributes.java:
```diff
@@ -251,6 +251,31 @@ public void clearObservable() {
 
+	/**
+	 * Marks the resource as only accessible using OSCORE.
+	 */
```
Please add:

```
@since 3.10
```

rikard-sics replied:
Done in commit e41bd07e2f4956c067eb4774a2edcee45e7e7a3a

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/server/resources/ResourceAttributes.java:
```diff
@@ -251,6 +251,31 @@ public void clearObservable() {
+	/**
+	 * Checks if the resource is only accessible using OSCORE.
+	 *
+	 * @return true, if only accessible using OSCORE
```
```
{@code true}
```

rikard-sics replied:
Added both in commit e41bd07e2f4956c067eb4774a2edcee45e7e7a3a

boaks commented on cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreResource.java:
```diff
@@ -34,6 +35,7 @@ public class OSCoreResource extends CoapResource {
 
 	public OSCoreResource(String name, final boolean isProtected) {
 		super(name);
+		getAttributes().setOscoreOnly();
```
Is there a relation betwenn "isProtected" and "Oscore Only"?
I searched for references to is Protected but could not find something.

rikard-sics replied:
Yes, great point, I missed that. At a minimum _setOscoreOnly()_ should only be invoked when _isProtected_ is set to true. Since _isProtected_ is the flag that is actually controlling if the resource is only accessible with OSCORE, it's not enough that the resource type is OSCoreResource.

I will take care of this.

rikard-sics replied:
Took care of only setting the OSCORE link target attribute if _isProtected_ is true in commit b9996e21d07b0c0bbe1ce45962ad54bf5cb03aff

rikard-sics commented:
Thanks for the feedback, I will take care of the points you mentioned.

rikard-sics commented:
I also took care of the issue of responding with RST to non-OSCORE requests to OSCORE-only resources in commit c1fb46aed9b8d684ffed895ee8da84bffe2924d5

If that looks good to you, we could keep it in this PR.

boaks commented on californium-core/src/main/java/org/eclipse/californium/core/server/resources/ResourceAttributes.java:
```diff
@@ -251,6 +251,34 @@ public void clearObservable() {
+	/**
+	 * Checks if the resource is only accessible using OSCORE.
+	 *
+	 * @since 3.10
```
please exchange the order of since and return.

rikard-sics replied:
I have fixed this also now

boaks commented:
Beside of the order of since and return, LGTM.
Please fix the order and squash the commits.

rikard-sics commented:
> Beside of the order of since and return, LGTM.
> Please fix the order and squash the commits.

Sounds great. I have now fixed the order and squashed the commits.
